### PR TITLE
Resolve intermittent `capture_output` deadlock 

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
We are continuing to see intermittent deadlocks or attempts to write to closed file descriptors in the HiGHS tests on main.  I think I have (finally) tracked down the root cause to the use of `capture_input()` in the *Gurobi* interface `__del__` method.  It looks like the GC will (randomly) interrupt a HiGHS test to run the collector (which, if it is collecting a Gurobi interface will call `capture_output`.  If this happens during another test's use of `capture_output.__enter__` or `.__exit__`, then the Gurobi interface's use of capture could happen when the underlying output streams are in a transitory state (and could be closed).

This PR resolves that by:
- rework the Gurobi interface to use it's own private Env (which suppresses the output to stdout and removes the need to use `capture_output` in `__del__`.
- Add thread locks so that one call to `capture_output` will not interrupts another fall to `__enter__ / `__exit__`
- improve some implementations within TeeStream / capture_output
- add additional tests that catch some edge cases identified during development

## Changes proposed in this PR:
- (see above)

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/common/tee.py [hunk 1]
```diff
@@ -667,15 +705,18 @@ def close(self, in_exception=False):
...
                 th.join(_poll)
+                _timeout += _poll
             self._threads[:] = [th for th in self._threads if th.is_alive()]
             if not self._threads:
                 break
-            _poll *= 2
-            if _poll_timeout <= _poll < 2 * _poll_timeout:
+            if _poll < _poll_timeout:
+                _poll *= 2.0
+            if _poll_timeout * 0.5 <= _poll < _poll_timeout:
```
Comment (mrmundt) [inline pyomo/common/tee.py:719]: Something I have wondered for a while - why the factors of 2? Is it arbitrarily decided?

File: pyomo/common/tests/test_tee.py [hunk 2]
```diff
@@ -559,6 +579,31 @@ def test_capture_output_stack_error(self):
             sys.stdout, sys.stderr = old
             logging.getLogger('pyomo.common.tee').handlers.clear()
 
+    def test_atomic_deadlock(self):
```
Comment (mrmundt) [inline pyomo/common/tests/test_tee.py:582]: This is a nifty test, I like it.

File: pyomo/contrib/solver/solvers/highs.py [hunk 3]
```diff
@@ -288,7 +288,7 @@ def set_instance(self, model):
                 f'({self.available()}).'
             )
 
-        with TeeStream(*ostreams) as t, capture_output(t.STDOUT, capture_fd=True):
+        with capture_output(TeeStream(*ostreams), capture_fd=True):
```
Comment (mrmundt) [inline pyomo/contrib/solver/solvers/highs.py:291]: What is the difference between doing this and just doing `with TeeStream(*ostreams) as OUT`? Do we have a preference? (I ask because we don't seem to have decided on a "standard" and we probably should)

Comment (jsiirola) [reply pyomo/contrib/solver/solvers/highs.py:291]: `capture_output` wraps the streams it is passed into a `TeeStream` in order to get handles it can give to stdout / stderr.  So the former implementation actually had 2 nested `TeeStream` objects.  If you just pass the `TeeStream` object to `capture_output`, it won't create a new `TeeStream`, so the new implementation only has a single `TeeStream`.

Comment (mrmundt) [reply pyomo/contrib/solver/solvers/highs.py:291]: Sorry, let me rephrase. In ipopt currently, we do this:
```
                with TeeStream(*ostreams) as t:
                    timer.start('subprocess')
                    process = subprocess.run(
                        cmd,
                        timeout=timeout,
                        env=env,
                        universal_newlines=True,
                        stdout=t.STDOUT,
                        stderr=t.STDERR,
                        check=False,
                    )
                    timer.stop('subprocess')
                    # This is the stuff we need to parse to get the iterations
                    # and time
                    (iters, ipopt_time_nofunc, ipopt_time_func, ipopt_total_time) = (
                        self._parse_ipopt_output(ostreams[0])
                    )
```

In GUROBI, we do:
```
            with capture_output(TeeStream(*ostreams), capture_fd=False):
                gurobi_model = gurobipy.Model()

                timer.start('transfer_model')
                x = gurobi_model.addMVar(
                    len(repn.columns),
                    lb=lb,
                    ub=ub,
                    obj=repn.c.todense()[0] if repn.c.shape[0] else 0,
                    vtype=vtype,
                )
```

Which one is preferable?

Comment (jsiirola) [reply pyomo/contrib/solver/solvers/highs.py:291]: They are both correct for their respective use cases:
- `subprocess.run` needs to be passed 2 TextIO objects, so it is appropriate (and necessary) to pass the raw TextIO objects from `TeeStream`.
- `capture_output` needs a single stream to send all output to (we have no added support for separate our/err targets (so internally it creates a TeeStream so that it can get raw file descriptors if it needs them (for `capture_fd=True`).  Here Passing a TeeStream directly avoids the need to create a second one that wraps the stream from the first.

Comment (mrmundt) [reply pyomo/contrib/solver/solvers/highs.py:291]: Got it, thank you!

**Review by mrmundt — APPROVED** [review]

**Comment by mrmundt** [comment]
Per @blnicho - we could merge this today if no one has time to do another review. Ergo, risking it for the biscuit and merging.
